nntrainer/graph/network_graph.cpp

```diff
@@ -457,6 +457,53 @@ void NetworkGraph::setBatchSize(unsigned int batch_size) {
     allocateTensors(exec_mode);
 }
 
+void NetworkGraph::tryZeroGrad(LayerNode *node) {
+  auto &rc = node->getRunContext();
+  auto current_execution_order = std::get<1>(node->getExecutionOrder());
+  auto num_weight = rc.getNumWeights();
+  for (unsigned i = 0; i < num_weight; ++i) {
+    if (!rc.weightHasGradient(i)) {
+      continue;
+    }
+
+    auto &grad = rc.getWeightGrad(i);
+    if (tensor_manager->isFirstAccess(grad.getName(),
+                                      current_execution_order)) {
+      grad.initialize();
+    }
+  }
+
+  auto num_input = rc.getNumInputs();
+  for (unsigned i = 0; i < num_input; ++i) {
+    auto deriv = rc.getOutgoingDerivative(i);
+    if (tensor_manager->isFirstAccess(deriv.getName(),
+                                      current_execution_order)) {
+      deriv.initialize();
+    }
+  }
+
+  /// @todo add loop for the trainable custom tensor
+}
+
+void NetworkGraph::tryApplyGradients(LayerNode *node,
+                                     std::function<void(Weight &)> apply_func) {
+  auto &rc = node->getRunContext();
+  auto current_execution_order = std::get<2>(node->getExecutionOrder());
+  auto num_weight = rc.getNumWeights();
+  for (unsigned i = 0; i < num_weight; ++i) {
+    if (!rc.weightHasGradient(i)) {
+      continue;
+    }
+
+    auto &grad = rc.getWeightGrad(i);
+    if (tensor_manager->isLastAccess(grad.getName(), current_execution_order)) {
+      apply_func(rc.getWeightObject(i));
+    }
+  }
+
+  /// @todo add loop for the trainable tensor as well
```

Comment:
no, this should not be done. trainable tensors are just given the gradients, but not applied. As of now, user is expected to apply the gradient itself.

Reply:
If it's current behavior, we won't going to need this, I'll delete

Reply:
yes.

Reply:
Done